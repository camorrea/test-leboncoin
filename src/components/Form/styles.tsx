import styled from 'styled-components'

export const Container = styled.div`
  height: 10rem;
  padding: 0.85rem;
`

export const ActionsContainer = styled.div`
  display: flex;
  justify-content: space-between;
`

export const Title = styled.h1`
  margin-bottom: 0.85rem;
  font-weight: bold;
  font-size: 0.9rem;
  line-height: 1rem;
  color: #333333;
`
