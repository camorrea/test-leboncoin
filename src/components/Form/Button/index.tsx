import { StyledButton } from './styles'

export const Button = StyledButton
