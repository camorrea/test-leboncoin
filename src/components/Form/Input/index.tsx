import React from 'react'
import { StyledInput } from './styles'

const Input = StyledInput

export default React.memo(Input)
