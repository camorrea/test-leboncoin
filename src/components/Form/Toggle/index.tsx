import React, { useCallback } from 'react'
import { HiddenLabel } from '../styles'
import {
  Background,
  Container,
  HiddenCheckbox,
  StyledToggleLabel,
  StyledToggleContainer,
  Switch
} from './styles'

type Props = React.InputHTMLAttributes<HTMLInputElement> & {
  checked?: boolean
  onChange?: () => void
  disabled?: boolean
  label: string
  id: string
}

export const ToggleContainer = StyledToggleContainer

export const ToggleLabel = StyledToggleLabel

const Toggle = (props: Props) => {
  const { checked, onChange, disabled, label, id } = props

  const handleChange = useCallback(() => {
    if (onChange && !disabled) onChange()
  }, [onChange, disabled])

  return (
    <Container>
      <HiddenLabel htmlFor={id}>{label}</HiddenLabel>
      <Background checked={checked} disabled={disabled}>
        <Switch checked={checked} disabled={disabled} />
      </Background>
      <HiddenCheckbox
        checked={checked}
        disabled={disabled}
        onChange={handleChange}
        id={id}
      />
    </Container>
  )
}

export default React.memo(Toggle)
