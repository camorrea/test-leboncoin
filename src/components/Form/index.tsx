import React, { useCallback, useRef, useState } from 'react'
import { Confidentiality } from '../../types'
import { Button } from './Button'
import Input from './Input'
import Toggle, { ToggleContainer, ToggleLabel } from './Toggle'
import { ActionsContainer, Container, HiddenLabel, Title } from './styles'

const Form = ({ postMessage }: { postMessage: any }) => {
  const postMessageInput = useRef<HTMLInputElement>()
  const [message, setMessage] = useState<string>('')
  const [isPrivate, setIsPrivate] = useState<boolean>(false)

  const handleSetMessage = useCallback(
    (event) => {
      if (message !== event.target.value) setMessage(event.target.value)
    },
    [message]
  )

  const handleToggleConfidentiality = useCallback(() => {
    setIsPrivate(!isPrivate)
  }, [isPrivate])

  const handlePostMessage = useCallback(
    (event) => {
      event.preventDefault()
      postMessage(message, isPrivate)
      setMessage('')
    },
    [message, isPrivate]
  )

  return (
    <Container onSubmit={handlePostMessage}>
      <Title>Post message</Title>
      <HiddenLabel htmlFor="messageField">Write your message here</HiddenLabel>
      <Input
        type="text"
        autoFocus={true}
        autoComplete="off"
        ref={postMessageInput}
        disabled={false}
        name="message"
        placeholder="Write your message here"
        value={message}
        onChange={handleSetMessage}
        id="messageField"
      />
      <ActionsContainer>
        <ToggleContainer>
          <Toggle
            checked={isPrivate}
            onChange={handleToggleConfidentiality}
            id="isPrivate"
            label={Confidentiality.private}
          />
          <ToggleLabel>{Confidentiality.private}</ToggleLabel>
        </ToggleContainer>
        <Button onClick={handlePostMessage} disabled={message === ''}>
          Send
        </Button>
      </ActionsContainer>
    </Container>
  )
}

export default React.memo(Form)
