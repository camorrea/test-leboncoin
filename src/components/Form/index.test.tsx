import { configure, shallow } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import React from 'react'
import renderer from 'react-test-renderer'

import Form from './'

configure({ adapter: new Adapter() })

const component = <Form onSubmit={jest.fn()} />

describe('Form Component', () => {
  it('should render properly', () => {
    const tree = renderer.create(component).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('should render a message input', () => {
    const wrapper = shallow(component)
    expect(wrapper.find('[data-test="messageInput"]').exists()).toBe(true)
  })

  it('should render an empty message input', () => {
    const wrapper = shallow(component)
    expect(wrapper.find('[data-test="messageInput"]').prop('value')).toBe('')
  })

  it('should render a toggle checkbox input', () => {
    const wrapper = shallow(component)
    expect(wrapper.find('[data-test="confidentialityToggle"]').exists()).toBe(
      true
    )
  })

  it('should render a toggle unchecked input', () => {
    const wrapper = shallow(component)
    expect(
      wrapper.find('[data-test="confidentialityToggle"]').prop('checked')
    ).toBe(false)
  })

  it('should render a submit button', () => {
    const wrapper = shallow(component)
    expect(wrapper.find('[data-test="submitButton"]').exists()).toBe(true)
  })

  it('a submit function should be triggered on click event', () => {
    const fn = jest.fn()
    const wrapper = shallow(<Form onSubmit={fn} />)
    wrapper.find('[data-test="submitButton"]').simulate('click')
    expect(fn).toHaveBeenCalled()
  })
})
