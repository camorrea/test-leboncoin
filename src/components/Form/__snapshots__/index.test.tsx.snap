// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form Component should render properly 1`] = `
<form
  className="sc-dlfnbm cpkrBB"
  onSubmit={[Function]}
>
  <h1
    className="sc-eCssSg bghaWm"
  >
    Post message
  </h1>
  <label
    className="sc-jSgupP kCrMOB"
    htmlFor="messageField"
  >
    Write your message here
  </label>
  <input
    autoComplete="off"
    autoFocus={true}
    className="sc-gsTCUz ljVjLm"
    data-test="messageInput"
    disabled={false}
    id="messageField"
    name="message"
    onChange={[Function]}
    placeholder="Write your message here"
    type="text"
    value=""
  />
  <div
    className="sc-hKgILt ezlxUD"
  >
    <div
      className="sc-gKsewC jmlHni"
    >
      <div
        className="sc-fubCfw caNQDg"
      >
        <label
          className="sc-jSgupP kCrMOB"
          htmlFor="isPrivate"
        >
          Private
        </label>
        <div
          checked={false}
          className="sc-pFZIQ fcZdPu"
        >
          <div
            checked={false}
            className="sc-jrAGrp imKRAw"
          />
        </div>
        <input
          checked={false}
          className="sc-kEjbxe hXwaRS"
          id="isPrivate"
          onChange={[Function]}
          type="checkbox"
        />
      </div>
      <div
        className="sc-iBPRYJ JXzPi"
      >
        Private
      </div>
    </div>
    <button
      className="sc-bdfBwQ kRDaXO"
      data-test="submitButton"
      disabled={true}
      onClick={[Function]}
    >
      Send
    </button>
  </div>
</form>
`;
